src/sidebar/components/test/user-menu-test.js

```diff
@@ -0,0 +1,268 @@
+'use strict';
+
+const { createElement } = require('preact');
+const { shallow } = require('enzyme');
+
+const UserMenu = require('../user-menu');
+const MenuItem = require('../menu-item');
+
+describe('UserMenu', () => {
+  let fakeAuth,
+    fakeBridge,
+    fakeInSidebar,
+    fakeIsThirdPartyUser,
+    fakeOnLogout,
+    fakeProfileBridgeEvent,
+    fakeServiceConfig,
+    fakeServiceUrl,
+    fakeSettings;
+
+  const createUserMenu = () => {
+    return shallow(
+      <UserMenu
+        auth={fakeAuth}
+        bridge={fakeBridge}
+        inSidebar={fakeInSidebar}
+        onLogout={fakeOnLogout}
+        serviceUrl={fakeServiceUrl}
+        settings={fakeSettings}
+      />
+    ).dive(); // Dive needed because this component uses `withServices`
+  };
+
+  const findMenuItem = (wrapper, labelText) => {
+    return wrapper
+      .find(MenuItem)
+      .filterWhere(n => n.prop('label') === labelText);
+  };
+
+  beforeEach(() => {
+    fakeAuth = {
+      displayName: 'Eleanor Fishtail',
+      status: 'logged-in',
+      userid: 'acct:eleanorFishtail@hypothes.is',
+      username: 'eleanorFishy',
+    };
+    fakeBridge = { call: sinon.stub() };
+    fakeInSidebar = true;
+    fakeIsThirdPartyUser = sinon.stub();
+    fakeOnLogout = sinon.stub();
+    fakeProfileBridgeEvent = 'profile-requested';
+    fakeServiceConfig = sinon.stub();
+    fakeServiceUrl = sinon.stub();
+    fakeSettings = {
+      authDomain: 'hypothes.is',
+    };
+
+    UserMenu.$imports.$mock({
```

Comment:
Originally I had some of this mocking and stubbing in `before` functions for nested tests (i.e. not at the top level) but I thought it might be easier to read if most of the test setup is in one place.

Reply:
This is fine. You can call `$imports.$mock` multiple times if you wish, so a test case or inner `beforeEach` can augment or replace mocks set by outer blocks. A pattern I've been using is to set up "useful defaults" in a `beforeEach` at the top of the test module and then more specific mocks in other `describe` or `it` blocks.